Allow embedders to open links in a new tab

webbeef opened the pull request:
This changes starts tracking the keyboard modifier state in the
`Constellation` and forwards it with every input event. The state
is used to modify the target of link click so when the
platform-dependent alternate action key is enabled, the target is
overriden to "_blank".

In addition, specification step numbers and text is updated.

Co-authored-by: Martin Robinson <mrobinson@igalia.com>
Signed-off-by: webbeef <me@webbeef.org>
Signed-off-by: Martin Robinson <mrobinson@igalia.com>

---

- [X] `./mach build -d` does not report any errors
- [X] `./mach test-tidy` does not report any errors
- [X] There are no tests for this change. This requires tests of the `WebView` interface which we don't support yet.

mrobinson commented:
I think this is handled internally in WebKit. For example, via the `create` signal in WebKitGTK: https://webkitgtk.org/reference/webkitgtk/stable/signal.WebView.create.html.

mrobinson commented:
Here's another signal involved in this decision: https://webkitgtk.org/reference/webkit2gtk/2.5.1/WebKitWebView.html#WebKitWebView-decide-policy

mrobinson commented on ports/servoshell/egl/android.rs:
```diff
@@ -557,8 +558,14 @@ impl HostTrait for HostCallbacks {
         match allow {
-            Ok(allow) => return allow.z().unwrap(),
-            Err(_) => return true,
+            Ok(allow) => {
+                return if allow.z().unwrap() {
+                    AllowNavigationOption::AllowInSameTab
+                } else {
+                    AllowNavigationOption::Disallow
+                }
+            },
+            Err(_) => return AllowNavigationOption::AllowInSameTab,
```
```suggestion
            let allow = allow.map_or(true, |allow| if allow.z().unwrap());
            match allow {
                true => AllowNavigationOption::AllowInSameTab,
                false => AllowNavigationOption::Disallow
            }
```

mrobinson reviewed (changes requested):
I think this should be handled internally in the Servo click handler. It doesn't make sense that every single embedder must specifically handle ctrl and meta clicks in order to detect when to open a new browser. In addition, in order to properly handle context menus Servo will need this support anyway. Instead, when the event handler detects a click or other action that wants to open a new window it should send a message to the embedder asking whether or not a new webview should be spawned.

mrobinson reviewed (changes requested):
Thanks! This is a bit closer, but one request for the design:

webbeef committed:
script: Allow opening links in a new `WebView`

This changes starts tracking the keyboard modifier state in the
`Constellation` and forwards it with every input event. The state
is used to modify the target of link click so when the
platform-dependent alternate action key is enabled, the target is
overriden to "_blank".

In addition, specification step numbers and text is updated.

Co-authored-by: Martin Robinson <mrobinson@igalia.com>
Signed-off-by: webbeef <me@webbeef.org>
Signed-off-by: Martin Robinson <mrobinson@igalia.com>

mrobinson commented:
@webbeef Okay, I've updated the PR with the changes I had in mind. Let me know if they look okay to you. Thanks!

mrobinson commented:
I think this deserves and outside review though.

webbeef reviewed (commented):
lgtm

jdm reviewed (approved):
This looks reasonable to me!
